src/ngx_http_lua_semaphore.c

```diff
@@ -0,0 +1,433 @@
+
+         dd("ngx_http_lua_free_sem sem_count:%lld",
+                 (long long int) ngx_http_lua_sem_free_count);
+    }
+}
+
+
+static ngx_int_t
+ngx_http_lua_semaphore_resume(ngx_http_request_t *r)
+{
+    lua_State                   *vm;
+    ngx_connection_t            *c;
+    ngx_int_t                   rc;
+    ngx_http_lua_ctx_t          *ctx;
+
+    ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
+    if (ctx == NULL) {
+        return NGX_ERROR;
+    }
+
+    ctx->resume_handler = ngx_http_lua_wev_handler;
+
+    c = r->connection;
+    vm = ngx_http_lua_get_lua_vm(r, ctx);
+
+    if (ctx->cur_co_ctx->sem_resume_status == 0) {
+         lua_pushboolean(ctx->cur_co_ctx->co, 1);
+         lua_pushnil(ctx->cur_co_ctx->co);
+
+    } else {
+         lua_pushboolean(ctx->cur_co_ctx->co, 0);
+         lua_pushstring(ctx->cur_co_ctx->co, "timeout");
+    }
+
+    rc = ngx_http_lua_run_thread(vm, r, ctx, 2);
+
+    ngx_log_debug1(NGX_LOG_DEBUG_HTTP, r->connection->log, 0,
+                "lua run thread returned %d", rc);
+
+    if (rc == NGX_AGAIN) {
+        return ngx_http_lua_run_posted_threads(c, vm, r, ctx);
+    }
+
+    if (rc == NGX_DONE) {
+        ngx_http_lua_finalize_request(r, NGX_DONE);
+        return ngx_http_lua_run_posted_threads(c, vm, r, ctx);
+    }
+
+    /* rc == NGX_ERROR || rc >= NGX_OK */
+    if (ctx->entered_content_phase) {
+        ngx_http_lua_finalize_request(r, rc);
+        return NGX_DONE;
+    }
+
+     return rc;
+}
+
+
+int
+ngx_http_lua_ffi_sem_new(ngx_http_lua_semaphore_t **psem, int n, char **errstr)
+{
+    ngx_http_lua_semaphore_t *sem;
+
+    sem = ngx_http_lua_alloc_sem(ngx_cycle->log);
+    if (sem == NULL) {
+        *errstr = "ngx_http_lua_ffi_sem_new ngx_alloc failed";
+        return NGX_ERROR;
+    }
+
+    ngx_memzero(sem, sizeof(ngx_http_lua_semaphore_t));
+    ngx_queue_init(&sem->wait_queue);
+
+    sem->count = n;
+    sem->wait_count = 0;
+    sem->in_post_event = 0;
+    sem->log = ngx_cycle->log;
+    *psem = sem;
+
+    dd("ngx_http_lua_ffi_sem_new semaphore %p value %d", sem, sem->count);
+
+    return NGX_OK;
+}
+
+#if 0
+int ngx_http_lua_ffi_sem_get_value(ngx_http_lua_semaphore_t *sem,char **err,size_t* errlen)
+{
+    if (sem == NULL )
+    {
+        return 
+    }
+
+    return sem->count - sem->wait_count;
+}
+#endif
+
+
+int
+ngx_http_lua_ffi_sem_post(ngx_http_lua_semaphore_t *sem, int n, char **errstr)
+{
+    if (sem == NULL) {
+        *errstr = "semaphore is null";
+        return NGX_ERROR;
+    }
+
+    sem->count += n;
+
+    dd("ngx_http_lua_ffi_sem_post semaphore %p value %d", sem, sem->count);
+
+    if (!sem->in_post_event && !ngx_queue_empty(&sem->wait_queue)) {
+
+        sem->sem_event.handler = ngx_http_lua_ffi_semaphore_handler;
+        sem->sem_event.data = sem;
+        sem->sem_event.log = sem->log;
+
+        sem->in_post_event = 1;
+        ngx_post_event((&sem->sem_event), &ngx_posted_events);
+    }
+
+    return NGX_OK;
+}
+
+
+int
```

Comment:
Okay, never mind, my bad. This is part of the ABI, so use of `int` here is appropriate.